File: answer_list/week30/3021-cg.go
package main

import "fmt"

type matrix [][]int

func main() {
	// Setup
	m0 := matrix{{0, 2, 1}, {3, 0, 5}, {1, 0, 3}, {1, 4, 1}}
	m1 := matrix{{1, 7, 8}, {3, 6, 9}, {1, 10, 2}, {7, 4, 1}}
	m2 := matrix{{1, 1, 0, 1}, {1, 1, 2, 5}, {2, 1, 6, 4}, {0, 1, 1, 3}}
	m3 := matrix{{1, 1, 0, 1}, {1, 1, 2, 6}, {2, 1, 6, 4}, {0, 1, 0, 3}}
	matrices := []matrix{m0, m1, m2, m3}
	chosenMatrixIndex := 1

	// Run
	max := 0
	fmt.Printf("Testing the following matrix: %d\n", matrices[chosenMatrixIndex])
	fmt.Printf("The max sum is: %d\n", maxMatrixSum(&matrices[chosenMatrixIndex], &max, 0, 0))
}

func maxMatrixSum(m *matrix, max *int, i int, j int) int {

	// basecase
	if i > len(*m)-1 ||
		j > len((*m)[i])-1 {
		return 0
	}

	// create incremented i, j
	iN := i + 1
	jN := j + 1

	// Recurse to find max of subtrees
	maxLeft := maxMatrixSum(m, max, i, jN)
	maxRight := maxMatrixSum(m, max, iN, j)

	// take the larger of the two
	maxOfAll := Max(maxLeft, maxRight) + (*m)[i][j]

	// reset max if needed
	if maxOfAll > *max {
		*max = maxOfAll
	}

	return maxOfAll
}

// Max util
func Max(x int, y int) int {
	if x > y {
		return x
	}
	return y
}

File: answer_list/2021/week30/3021-cg.cpp

/*
Weekly Coding Problem

Week 30, 2021

This question was asked by Zillow.

You are given a `2-d` matrix where each cell represents number of coins in that cell. Assuming we start at `matrix[0][0]`, and can only move right or down, find the maximum number of coins you can collect by the bottom right corner.

For example, in this matrix:

0 3 1 1
2 0 0 4
1 5 3 1

The most we can collect is `0 + 2 + 1 + 5 + 3 + 1 = 12` coins.

 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <cstdlib>
#include <iomanip>

typedef std::vector<std::vector<int>> matrix;
int maxMatrixSum(matrix &, int &, int, int);
void printMatrix(matrix &, std::ostream &);

int main()
{
  matrix m0{{0, 2, 1}, {3, 0, 5}, {1, 0, 3}, {1, 4, 1}};
  matrix m1{{1, 7, 8}, {3, 6, 9}, {1, 10, 2}, {7, 4, 1}};
  matrix m2{{1, 1, 0, 1}, {1, 1, 2, 5}, {2, 1, 6, 4}, {0, 1, 1, 3}};
  matrix m3{{1, 1, 0, 1}, {1, 1, 2, 6}, {2, 1, 6, 4}, {0, 1, 0, 3}};
  int max = 0;

  std::vector<matrix> v{m0, m1, m2, m3};
  std::cout << "\nChoose one of the following matrices:\n";
  for (int i = 0; i < v.size(); ++i)
  {
    std::cout << "Matrix: " << i << "\n";
    printMatrix(v[i], std::cout);
  }

  int choice;
  std::cout << "Choose a matrix number (0-" << v.size() - 1 << "): ";
  std::cin >> choice;
  std::cout << "Testing Matrix: " << choice << '\n';
  printMatrix(v[choice], std::cout);
  std::cout << "Max sum: " << maxMatrixSum(v[choice], max, 0, 0) << '\n';
  return 0;
}

// Solution
int maxMatrixSum(matrix &m, int &max, int i, int j)
{
  if (i > m.size() - 1 || j > m[i].size() - 1)
    return 0;

  int iN = i + 1;
  int jN = j + 1;

  // Get max of subtrees
  int maxLeft = maxMatrixSum(m, max, i, jN);
  int maxRight = maxMatrixSum(m, max, iN, j);

  int maxOfAll = std::max(maxLeft, maxRight) + m[i][j];

  if (maxOfAll > max)
    max = maxOfAll;

  return maxOfAll;
}

// Print utility
void printMatrix(matrix &m, std::ostream &os)
{
  for (int i = 0; i < m[0].size(); ++i)
  {
    for (int j = 0; j < m.size(); ++j)
    {
      os << std::setw(3) << m[j][i];
      if (j == m.size() - 1)
        os << '\n';
      else
        os << ' ';
    }
  }
}
